File: lib/sp500/market-calendar.ts
/**
 * US equity session helpers for S&P 500 0DTE / weekly markets (MVP).
 *
 * Official Alpaca docs (pricing only — no free-tier calendar API):
 * - https://alpaca.markets/docs/
 * - https://alpaca.markets/docs/api-references/market-data-api/
 */

const NYSE_HOLIDAYS = new Set<string>([
  "2025-01-01",
  "2025-01-20",
  "2025-02-17",
  "2025-04-18",
  "2025-05-26",
  "2025-06-19",
  "2025-07-04",
  "2025-09-01",
  "2025-11-27",
  "2025-12-25",
  "2026-01-01",
  "2026-01-19",
  "2026-02-16",
  "2026-04-03",
  "2026-05-25",
  "2026-06-19",
  "2026-07-03",
  "2026-09-07",
  "2026-11-26",
  "2026-12-25",
  "2027-01-01",
  "2027-01-18",
  "2027-02-15",
  "2027-03-26",
  "2027-05-31",
  "2027-06-18",
  "2027-07-05",
  "2027-09-06",
  "2027-11-25",
  "2027-12-24",
]);

function pad(n: number): string {
  return String(n).padStart(2, "0");
}

export function isoDateUtcParts(y: number, m: number, day: number): string {
  return `${y}-${pad(m)}-${pad(day)}`;
}

/** Calendar date in America/New_York as a UTC-midnight Date + ISO string. */
export function todayEt(now = new Date()): { date: Date; iso: string } {
  const fmt = new Intl.DateTimeFormat("en-CA", {
    timeZone: "America/New_York",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  });
  const parts = fmt.formatToParts(now);
  const y = Number(parts.find((p) => p.type === "year")?.value);
  const m = Number(parts.find((p) => p.type === "month")?.value);
  const day = Number(parts.find((p) => p.type === "day")?.value);
  return { date: new Date(Date.UTC(y, m - 1, day)), iso: isoDateUtcParts(y, m, day) };
}

export function isWeekend(day: Date): boolean {
  // UTC-midnight Date representing an ET calendar day
  const wd = day.getUTCDay();
  return wd === 0 || wd === 6;
}

export function isNyseHoliday(day: Date): boolean {
  const iso = isoDateUtcParts(day.getUTCFullYear(), day.getUTCMonth() + 1, day.getUTCDate());
  return NYSE_HOLIDAYS.has(iso);
}

export function isTradingDay(day: Date): boolean {
  return !isWeekend(day) && !isNyseHoliday(day);
}

export function nextTradingDay(onOrAfter: Date): Date {
  const d = new Date(onOrAfter);
  for (let i = 0; i < 14; i++) {
    if (isTradingDay(d)) return d;
    d.setUTCDate(d.getUTCDate() + 1);
  }
  return onOrAfter;
}

/** Friday on/after date, rolled forward if that Friday is a holiday. */
export function nextWeeklyExpiration(onOrAfter: Date): Date {
  const d = new Date(onOrAfter);
  d.setUTCHours(0, 0, 0, 0);
  const delta = (5 - d.getUTCDay() + 7) % 7;
  d.setUTCDate(d.getUTCDate() + delta);
  return nextTradingDay(d);
}

/** 16:00 America/New_York on the expiration calendar day (handles EST/EDT). */
export function sessionCloseMs(expiration: Date): number {
  const y = expiration.getUTCFullYear();
  const month = expiration.getUTCMonth();
  const day = expiration.getUTCDate();
  for (const utcHour of [20, 21]) {
    const ms = Date.UTC(y, month, day, utcHour, 0, 0);
    const parts = new Intl.DateTimeFormat("en-GB", {
      timeZone: "America/New_York",
      hour: "2-digit",
      minute: "2-digit",
      day: "2-digit",
      hour12: false,
    }).formatToParts(new Date(ms));
    const hour = Number(parts.find((p) => p.type === "hour")?.value);
    const etDay = Number(parts.find((p) => p.type === "day")?.value);
    if (hour === 16 && etDay === day) return ms;
  }
  return Date.UTC(y, month, day, 20, 0, 0);
}

export type SessionPhase = "pre_market" | "regular" | "after_hours" | "closed";

export function sessionPhase(now = new Date()): SessionPhase {
  const { date, iso } = todayEt(now);
  if (!isTradingDay(date) || NYSE_HOLIDAYS.has(iso)) return "closed";

  const parts = new Intl.DateTimeFormat("en-GB", {
    timeZone: "America/New_York",
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  }).formatToParts(now);
  const hour = Number(parts.find((p) => p.type === "hour")?.value);
  const minute = Number(parts.find((p) => p.type === "minute")?.value);
  const mins = hour * 60 + minute;
  if (mins < 9 * 60 + 30) return "pre_market";
  if (mins < 16 * 60) return "regular";
  return "after_hours";
}

File: backend/integrations/alpaca/market_calendar.py
"""US equity session helpers for S&P 500 market generation / resolution.

Handles weekends, common NYSE holidays, and coarse after-hours detection so
0DTE / weekly markets are not opened or resolved on non-sessions.

Alpaca Market Data does not publish a session calendar on the free tier, so this
module ships a static NYSE holiday set sufficient for the MVP.
"""

from __future__ import annotations

from datetime import date, datetime, time, timedelta
from zoneinfo import ZoneInfo

_ET = ZoneInfo("America/New_York")

# Observed NYSE full-day closures (extend annually for production calendars).
# Early-close days are treated as regular sessions for MVP strike generation;
# resolution still waits for the official daily bar.
_NYSE_HOLIDAYS: frozenset[date] = frozenset(
    {
        # 2025
        date(2025, 1, 1),
        date(2025, 1, 20),
        date(2025, 2, 17),
        date(2025, 4, 18),
        date(2025, 5, 26),
        date(2025, 6, 19),
        date(2025, 7, 4),
        date(2025, 9, 1),
        date(2025, 11, 27),
        date(2025, 12, 25),
        # 2026
        date(2026, 1, 1),
        date(2026, 1, 19),
        date(2026, 2, 16),
        date(2026, 4, 3),
        date(2026, 5, 25),
        date(2026, 6, 19),
        date(2026, 7, 3),  # Independence Day observed
        date(2026, 9, 7),
        date(2026, 11, 26),
        date(2026, 12, 25),
        # 2027
        date(2027, 1, 1),
        date(2027, 1, 18),
        date(2027, 2, 15),
        date(2027, 3, 26),
        date(2027, 5, 31),
        date(2027, 6, 18),  # Juneteenth observed
        date(2027, 7, 5),  # Independence Day observed
        date(2027, 9, 6),
        date(2027, 11, 25),
        date(2027, 12, 24),  # Christmas observed
    }
)

_REGULAR_OPEN = time(9, 30)
_REGULAR_CLOSE = time(16, 0)


def us_equity_now(now: datetime | None = None) -> datetime:
    current = now or datetime.now(tz=_ET)
    if current.tzinfo is None:
        return current.replace(tzinfo=_ET).astimezone(_ET)
    return current.astimezone(_ET)


def us_equity_today(now: datetime | None = None) -> date:
    return us_equity_now(now).date()


def is_weekend(day: date) -> bool:
    return day.weekday() >= 5


def is_nyse_holiday(day: date) -> bool:
    return day in _NYSE_HOLIDAYS


def is_trading_day(day: date) -> bool:
    """True when the US equity cash session is scheduled to open."""
    return not is_weekend(day) and not is_nyse_holiday(day)


def next_trading_day(on_or_after: date) -> date:
    day = on_or_after
    for _ in range(14):
        if is_trading_day(day):
            return day
        day += timedelta(days=1)
    return on_or_after


def previous_trading_day(on_or_before: date) -> date:
    day = on_or_before
    for _ in range(14):
        if is_trading_day(day):
            return day
        day -= timedelta(days=1)
    return on_or_before


def next_weekly_expiration(on_or_after: date) -> date:
    """Friday-or-next-trading-day weekly close (skips holiday Fridays)."""
    # weekday(): Mon=0 … Fri=4
    delta = (4 - on_or_after.weekday()) % 7
    friday = on_or_after + timedelta(days=delta)
    return next_trading_day(friday)


def session_phase(now: datetime | None = None) -> str:
    """Return ``pre_market`` | ``regular`` | ``after_hours`` | ``closed``."""
    current = us_equity_now(now)
    day = current.date()
    if not is_trading_day(day):
        return "closed"
    clock = current.timetz().replace(tzinfo=None)
    if clock < _REGULAR_OPEN:
        return "pre_market"
    if clock < _REGULAR_CLOSE:
        return "regular"
    return "after_hours"


def is_regular_session(now: datetime | None = None) -> bool:
    return session_phase(now) == "regular"


def quote_is_stale_or_thin(
    *,
    has_latest_trade: bool,
    has_daily_bar: bool,
    previous_close_only: bool,
    now: datetime | None = None,
) -> bool:
    """Heuristic for after-hours / low-volume / halt-like thin books."""
    phase = session_phase(now)
    if previous_close_only:
        return True
    if phase in {"pre_market", "after_hours", "closed"} and not has_latest_trade:
        return True
    if phase == "regular" and not has_latest_trade and not has_daily_bar:
        return True
    return False
